File: modules/captivet-durable-recorder/ios/CaptivetDurableRecorderModule.swift
import ExpoModulesCore

/// Expo Module bridge for the durable audio recorder (iOS).
///
/// Every AsyncFunction delegates to `DurableRecorderEngine.shared`, which owns
/// the AVAudioEngine capture pipeline, ADTS writer, manifest IO, and recovery
/// scan. This class only marshals arguments, wires events, and never traps: all
/// native failures surface as a typed `error` event + a rejected promise
/// carrying a `CODE: message` reason (see index.ts contract).
///
/// The synchronous `getLiveStats` powers the 500 ms foreground level-meter feed
/// and is intentionally NOT tied to the ~2 s durable commit cadence.
public class CaptivetDurableRecorderModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CaptivetDurableRecorder")

    Events("recordingProgress", "liveStats", "stateChanged", "interruption", "error")

    OnCreate {
      // Route native events to JS. sendEvent wants [String: Any?]; adapt here.
      DurableRecorderEngine.shared.eventEmitter = { [weak self] name, body in
        self?.sendEvent(name, body.mapValues { $0 as Any? })
      }
    }

    OnDestroy {
      DurableRecorderEngine.shared.detachEmitter()
    }

    // MARK: - Capture ops

    AsyncFunction("start") { (input: StartInput) -> [String: Any] in
      try DurableRecorderEngine.shared.start(
        userId: input.userId,
        slotId: input.slotId,
        recordingId: input.recordingId,
        commitIntervalMs: input.commitIntervalMs,
        sampleRate: input.sampleRate,
        bitrate: input.bitrate
      )
    }

    AsyncFunction("pause") { () -> [String: Any] in
      try DurableRecorderEngine.shared.pause()
    }

    AsyncFunction("resume") { (input: UserRecordingInput) -> [String: Any] in
      try DurableRecorderEngine.shared.resume(userId: input.userId, recordingId: input.recordingId)
    }

    // stop() may be called with no argument (see index.ts); accept an optional.
    AsyncFunction("stop") { (input: StopInput?) -> [String: Any] in
      try DurableRecorderEngine.shared.stop(userId: input?.userId, recordingId: input?.recordingId)
    }

    AsyncFunction("discard") { (input: UserRecordingInput) in
      try DurableRecorderEngine.shared.discard(userId: input.userId, recordingId: input.recordingId)
    }

    AsyncFunction("purgeAfterUpload") { (input: UserRecordingInput) in
      try DurableRecorderEngine.shared.purgeAfterUpload(userId: input.userId, recordingId: input.recordingId)
    }

    // MARK: - Manifest mutations (atomic temp+rename)

    AsyncFunction("setServerRecordingId") { (input: SetServerIdInput) in
      try DurableRecorderEngine.shared.setServerRecordingId(
        userId: input.userId,
        recordingId: input.recordingId,
        serverRecordingId: input.serverRecordingId
      )
    }

    AsyncFunction("setPendingConfirm") { (input: SetPendingConfirmInput) in
      try DurableRecorderEngine.shared.setPendingConfirm(
        userId: input.userId,
        recordingId: input.recordingId,
        pendingConfirmJson: input.pendingConfirmJson
      )
    }

    AsyncFunction("markUploaded") { (input: MarkUploadedInput) in
      try DurableRecorderEngine.shared.markUploaded(
        userId: input.userId,
        recordingId: input.recordingId,
        confirmedUploadAt: input.confirmedUploadAt
      )
    }

    // MARK: - Read / recovery ops (degrade to nil/[] on the JS side if absent)

    AsyncFunction("getStatus") { () -> [String: Any]? in
      DurableRecorderEngine.shared.getStatus()
    }

    AsyncFunction("getManifest") { (input: UserRecordingInput) -> [String: Any]? in
      try DurableRecorderEngine.shared.getManifest(userId: input.userId, recordingId: input.recordingId)
    }

    AsyncFunction("listRecoverableSessions") { (userId: String) -> [[String: Any]] in
      try DurableRecorderEngine.shared.listRecoverableSessions(userId: userId)
    }

    // MARK: - Synchronous live feed

    Function("getLiveStats") { () -> [String: Any]? in
      DurableRecorderEngine.shared.getLiveStats()
    }
  }
}

// MARK: - Typed exception

/// Rejects a promise with a stable JS error. The precise machine-readable code
/// travels on the `error` EVENT ({ code, message }); the rejection `message`
/// mirrors it as `CODE: message` so a caller can parse either channel.
internal final class DurableException: GenericException<String> {
  override var reason: String { param }
}

// MARK: - Input records

internal struct StartInput: Record {
  @Field var userId: String = ""
  @Field var slotId: String = ""
  @Field var recordingId: String = ""
  @Field var commitIntervalMs: Int? = nil
  @Field var sampleRate: Int? = nil
  @Field var bitrate: Int? = nil
}

internal struct UserRecordingInput: Record {
  @Field var userId: String = ""
  @Field var recordingId: String = ""
}

internal struct StopInput: Record {
  @Field var userId: String? = nil
  @Field var recordingId: String? = nil
}

internal struct SetServerIdInput: Record {
  @Field var userId: String = ""
  @Field var recordingId: String = ""
  @Field var serverRecordingId: String = ""
}

internal struct MarkUploadedInput: Record {
  @Field var userId: String = ""
  @Field var recordingId: String = ""
  @Field var confirmedUploadAt: String = ""
}

internal struct SetPendingConfirmInput: Record {
  @Field var userId: String = ""
  @Field var recordingId: String = ""
  @Field var pendingConfirmJson: String? = nil
}

File: modules/captivet-durable-recorder/android/src/main/java/expo/modules/captivetdurablerecorder/CaptivetDurableRecorderModule.kt
package expo.modules.captivetdurablerecorder

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo bridge for the durable recorder. It is a thin translation layer: it
 * validates the React context, forwards each JS call to the process-singleton
 * DurableRecorderEngine, and wires native events to sendEvent.
 *
 * The function names, argument/return shapes, and event names MUST match
 * modules/captivet-durable-recorder/index.ts (the JS contract). Every native
 * failure surfaces as a rejected promise carrying a typed DurableErrors code —
 * nothing here terminates the process (CLAUDE.md Rule 1).
 *
 * The capture pipeline lives in the engine (a singleton) rather than this module
 * so getStatus / getLiveStats / events keep working across a JS reload while
 * native keeps recording, and so the foreground service can reach the engine.
 */
class CaptivetDurableRecorderModule : Module() {
  private fun requireContext() = appContext.reactContext?.applicationContext
    ?: throw DurableRecorderException(DurableErrors.NO_CONTEXT, "React context unavailable")

  override fun definition() = ModuleDefinition {
    Name("CaptivetDurableRecorder")

    Events(
      "recordingProgress",
      "liveStats",
      "stateChanged",
      "interruption",
      "error",
    )

    OnCreate {
      // Route engine events through this module. Set on create so events flow
      // even before the first start()/resume() and after a JS reload.
      DurableRecorderEngine.attach { name, body -> sendEvent(name, body) }
    }

    // --- Capture ops (reject with a typed code so the hook can fall back) ---

    AsyncFunction("start") { input: Map<String, Any?> ->
      DurableRecorderEngine.start(requireContext(), input)
    }

    AsyncFunction("pause") {
      DurableRecorderEngine.pause()
    }

    AsyncFunction("resume") { input: Map<String, Any?> ->
      DurableRecorderEngine.resume(requireContext(), input)
    }

    AsyncFunction("stop") { input: Map<String, Any?>? ->
      DurableRecorderEngine.stop(requireContext(), input)
    }

    AsyncFunction("discard") { input: Map<String, Any?> ->
      DurableRecorderEngine.discard(requireContext(), input)
    }

    AsyncFunction("purgeAfterUpload") { input: Map<String, Any?> ->
      DurableRecorderEngine.purgeAfterUpload(requireContext(), input)
    }

    AsyncFunction("setServerRecordingId") { input: Map<String, Any?> ->
      DurableRecorderEngine.setServerRecordingId(requireContext(), input)
    }

    AsyncFunction("setPendingConfirm") { input: Map<String, Any?> ->
      DurableRecorderEngine.setPendingConfirm(requireContext(), input)
    }

    AsyncFunction("resetUploadAttempt") { input: Map<String, Any?> ->
      DurableRecorderEngine.resetUploadAttempt(requireContext(), input)
    }

    AsyncFunction("markUploaded") { input: Map<String, Any?> ->
      DurableRecorderEngine.markUploaded(requireContext(), input)
    }

    // --- Read / recovery ops (resolve to null / [] when nothing to return) ---

    AsyncFunction("getStatus") {
      DurableRecorderEngine.getStatus()
    }

    AsyncFunction("getManifest") { input: Map<String, Any?> ->
      DurableRecorderEngine.getManifest(requireContext(), input)
    }

    AsyncFunction("listRecoverableSessions") { userId: String ->
      DurableRecorderEngine.listRecoverableSessions(requireContext(), userId)
    }

    // Synchronous high-frequency live feed (the primary 500ms-poll path).
    Function("getLiveStats") {
      DurableRecorderEngine.getLiveStats()
    }

    OnDestroy {
      DurableRecorderEngine.detach()
    }
  }
}
